### PR TITLE
Better recursive parse

### LINKED ISSUE #8: Fix measure and classical reg
For now the number of classical regs and the memory slots for storing measurement values are hardcoded to a fixed value. This is not useful incase of larger/practical circuits

### LINKED ISSUE #10: Add unit tests

### LINKED ISSUE #13: Use gates outside basis_gates
In the step 2 in the Readme, while creating a `IBMQReg` the user is given a choice of backends. For each backend the interface lists out the number of qubits and basis_gates supported by the backend. 
Now if we try to run higher qubits or gates outside the basis_gates of any backend, we get a `ERROR_RUNNING_JOB` return value. 
For eg : `ibm_santiago` supports  `basis_gates=Any["id", "u1", "u2", "u3", "cx"]`. Now if my circuit contains the gate `rx`, it will return `ERROR_RUNNING_JOB`. 

Apparently one might conclude that it;s not possible to run certain gates on some backends. Surprisingly, rx, ry, rz gates are not supported by ibm_qasm_simulator too! But the qasm specification does list it here https://github.com/Qiskit/openqasm/blob/master/examples/generic/qelib1.inc#L38-L42. 

I wonder if we're missing something here? Do we need to represent those gates in terms of the supported gates for each backend(eg: rz interms of other gates if possible)? Or is the basis_gates list sort of misleading?

I'd really love to discuss this here before asking the IBMQ folks?
Any ideas @Roger-luo @GiggleLiu

### PR DESCRIPTION
**Pull request by GiggleLiu**
```julia
julia> using IBMQJulia, Yao

julia> IBMQJulia.generate_experiment(chain([control(4, 1, 2=>shift(3))]))
Dict{String,Any} with 3 entries:
  "instructions" => Any[Dict{String,Any}("name"=>"cu1","params"=>[3.0],"qubits"=>[0, 1])]
  "header"       => Dict{String,Any}("clbit_labels"=>Array{Any,1}[["c", 0], ["c", 1]],"qubit_labels"=>Array{Any,1}[["q", 0], ["q", 1], …
  "config"       => Dict{Any,Any}()

julia> IBMQJulia.generate_inst(put(10, (2,1)=>put(2, 2=>chain(X, Y, Z, Measure()))))
4-element Array{Any,1}:
 Dict{String,Any}("name" => "x","qubits" => [0])
 Dict{String,Any}("name" => "y","qubits" => [0])
 Dict{String,Any}("name" => "z","qubits" => [0])
 Dict{String,Any}("name" => "measure","memory" => [0.0],"qubits" => [0])
```

## Notice

* the indices now starts from 0, please check this is correct.
* the type parameter N in `AbstractionBlock{N}` means the number of qubits.

### CODE REVIEW

**Comment by Sov-trotter** [comment]
This is really cool. 
>the indices now starts from 0, please check this is correct.

So I tried to run the circuit and it works. So I think there isn't any issue here.

File: test/testqobj.jl [hunk 1]
```diff
@@ -50,12 +51,15 @@ end
...
+    c = build_circuit(4, 1, [1=>2, 2=>3, 3=>4])
+    inst = generate_inst(c)
+    c2 = inst |> inst2qbir
+    @test operator_fidelity(c, c2) ≈ 1
 end
 
 @testset "Variational Quantum Eigen Solver" begin
-    for n in 3:10
+    for n in [3, 6, 7, 8, 9]
         circuit = dispatch!(variational_circuit(n, n+1),:random)
```
Comment (GiggleLiu) [inline test/testqobj.jl:63]: We should test different gates one by one. Rather than testing circuits as a whole.

Comment (Sov-trotter) [reply test/testqobj.jl:63]: Yeah. I'll add those soon

Comment (GiggleLiu) [reply test/testqobj.jl:63]: perfect!

**Comment by GiggleLiu** [comment]
I just removed the API test first. Feel free to add it back after the mock interface is ready.

**Comment by Sov-trotter** [comment]
>Feel free to add it back after the mock interface is ready.

Can you elaborate on this?

**Comment by GiggleLiu** [comment]
> > Feel free to add it back after the mock interface is ready.
> 
> Can you elaborate on this?

Current api tests fails, so I commented it out. I think it is not likely a bug.

**Comment by Sov-trotter** [comment]
Looks good to go?
Should we register it and tag a release first?

**Comment by GiggleLiu** [comment]
Sounds Good!

**Pull request merged**